odoo/osv/expression.py

```diff
@@ -1323,10 +1310,10 @@ def push_result(sql):
             elif field.type == 'many2one':
                 if operator in HIERARCHY_FUNCS:
                     ids2 = to_ids(right, comodel, leaf)
-                    if field.comodel_name != model._name:
+                    if field.comodel_name != model._name:  # TODO: change the semantic
```

Comment:
Example: model is `sale.order` and domain is`[('partner_id', 'child_of', [1])]`, so comodel != model.